### PR TITLE
feat(destination): add dedup support for identify call in Braze

### PR DESCRIPTION
**Pull request by sauravlal15**
## Description of the change

> Here, it will add dedup support for identify call in Braze

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [✅] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/728)

### CODE REVIEW

File: integrations/Braze/browser.js [hunk 1]
```diff
@@ -28,7 +30,8 @@ class Braze {
     }
 
     this.name = NAME;
-
+    this.previousPayload = null;
+    this.supportDedup = true;
```
Comment (sauravlal15) [inline integrations/Braze/browser.js:34]: It will be taken from config, once it's reviewed I'll change it.

File: integrations/Braze/browser.js [hunk 2]
```diff
@@ -172,28 +147,121 @@ class Braze {
...
     ];
 
+    // deep clone the traits object
+    const traits = cloneDeep(rudderElement.message.context.traits);
+
     reserved.forEach((element) => {
       delete traits[element];
     });
+    if (this.supportDedup && this.previousPayload !== null) {
+      const prevUserId = get(this.previousPayload, "rudderElement.message.userId");
```
Comment (ItsSudip) [inline integrations/Braze/browser.js:166]: Does updating the userId, will create a new user? If true then, if the userId is updated apart from going with the new flow can't we go with the existing flow?

File: integrations/Braze/browser.js [hunk 3]
```diff
@@ -172,28 +147,121 @@ class Braze {
...
+        "rudderElement.message.context.traits.gender"
+      );
+      const prevLastname = get(
+        this.previousPayload,
+        "rudderElement.message.context.traits.lastname"
+      );
+      const prevPhone = get(
+        this.previousPayload,
+        "rudderElement.message.context.traits.phone"
+      );
```
Comment (ItsSudip) [inline integrations/Braze/browser.js:194]: Use optional chaining.

File: integrations/Braze/browser.js [hunk 4]
```diff
@@ -172,28 +147,121 @@ class Braze {
...
+      if (gender && gender !== prevGender) {
+        window.braze.getUser().setGender(this.formatGender(gender));
+      }
+      if (lastname && lastname !== prevLastname) {
+        window.braze.getUser().setLastName(lastname);
+      }
+      if (phone && phone !== prevPhone) {
+        window.braze.getUser().setPhoneNumber(phone);
+      }
+      // const prevtraits = cloneDeep();
```
Comment (ItsSudip) [inline integrations/Braze/browser.js:227]:
```suggestion
```

File: integrations/Braze/browser.js [hunk 5]
```diff
@@ -172,28 +147,121 @@ class Braze {
...
+      }
+      if (gender && gender !== prevGender) {
+        window.braze.getUser().setGender(this.formatGender(gender));
+      }
+      if (lastname && lastname !== prevLastname) {
+        window.braze.getUser().setLastName(lastname);
+      }
+      if (phone && phone !== prevPhone) {
+        window.braze.getUser().setPhoneNumber(phone);
+      }
```
Comment (ItsSudip) [inline integrations/Braze/browser.js:226]: This part is repeating, can't we store them these attributes and use?

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: integrations/Braze/browser.js (hunk 3 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js]: done

File: integrations/Braze/browser.js [hunk 6]
```diff
@@ -1,8 +1,10 @@
 /* eslint-disable class-methods-use-this */
+import cloneDeep from "lodash.clonedeep";
 import { del } from "obj-case";
 import logger from "../../utils/logUtil";
 import { LOAD_ORIGIN } from "../ScriptLoader";
 import { BrazeOperationString, NAME } from "./constants";
+import get from "get-value";
```
[the remark this answers is not in the document]
Comment (sauravlal15) [reply integrations/Braze/browser.js]: updated

File: integrations/Braze/browser.js [hunk 7]
```diff
@@ -172,28 +147,121 @@ class Braze {
...
       "home_city",
       "bio",
-      "gender",
-      "phone",
       "email_subscribe",
       "push_subscribe",
     ];
 
+    // deep clone the traits object
+    const traits = cloneDeep(rudderElement.message.context.traits);
```
[the remark this answers is not in the document]
Comment (sauravlal15) [reply integrations/Braze/browser.js]: updated

File: integrations/Braze/browser.js (hunk 2 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js]: yes, it will change, updated

File: integrations/Braze/browser.js (hunk 4 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js]: yeah, removed

File: integrations/Braze/browser.js (hunk 5 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js]: yes, updated

File: integrations/Braze/browser.js [hunk 8]
```diff
@@ -172,28 +154,103 @@ class Braze {
...
+      window.braze.getUser().setEmail(email);
+    }
+    // function set firstName
+    function setFirstName() {
+      window.braze.getUser().setFirstName(firstname);
+    }
+    // function set gender
+    function setGender(genderName) {
+      window.braze.getUser().setGender(genderName);
+    }
```
Comment (ItsSudip) [inline integrations/Braze/browser.js:191]: Why do we need to send the parameter here can't we use the gender?

Comment (sauravlal15) [reply integrations/Braze/browser.js:191]: Yes, we have access to gender, but we don't have access to the `this.formatGender()` inside `setGender()` function, so I passed it as a parameter.

File: integrations/Braze/browser.js [hunk 9]
```diff
@@ -161,17 +162,56 @@ class Braze {
...
+    function setLastName() {
+      window.braze.getUser().setLastName(lastname);
+    }
+    function setPhone() {
+      window.braze.getUser().setPhoneNumber(phone);
+    }
 
     // deep clone the traits object
-    const traits = cloneDeep(rudderElement.message.context.traits);
+    const traits = cloneDeep(rudderElement.message?.context?.traits);
```
[the remark this answers is not in the document]
Comment (sauravlal15) [reply integrations/Braze/browser.js:205]: updated

**Review by ItsSudip — DISMISSED** [review]

**Review by MoumitaM — APPROVED** [review]

File: integrations/Braze/browser.js [hunk 10]
```diff
@@ -172,28 +154,106 @@ class Braze {
...
+      const prevTraits = prevMessage?.context?.traits;
+      const prevAddress = prevTraits?.address;
+      const prevBirthday = prevTraits?.birthday || prevTraits?.dob;
+      const prevEmail = prevTraits?.email;
+      const prevFirstname = prevTraits?.firstname || prevTraits?.firstName;
+      const prevGender = prevTraits?.gender;
+      const prevLastname = prevTraits?.lastname || prevTraits?.lastName;
+      const prevPhone = prevTraits?.phone;
+
+      if (address && !isEqual(address, prevAddress)) setAddress();
```
Comment (krishna2020) [inline integrations/Braze/browser.js:226]: address is an object, we have to do check using custom comparator including all the address fields

File: integrations/Braze/browser.js [hunk 11]
```diff
@@ -1,4 +1,7 @@
+/* eslint-disable consistent-return */
```
Comment (krishna2020) [inline integrations/Braze/browser.js:1]: Add the relevant test cases for the new dedupe logic

Comment (sauravlal15) [reply integrations/Braze/browser.js:1]: Yeah, will update.

File: integrations/Braze/browser.js (hunk 10 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js:226]: Yes, we are using lodash isEqual which will deeply check all fields and return boolean accordingly.

File: integrations/Braze/browser.js (hunk 11 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js:1]: updated

File: integrations/Braze/browser.js [hunk 12]
```diff
@@ -1,4 +1,7 @@
...
 /* eslint-disable class-methods-use-this */
+import cloneDeep from "lodash.clonedeep";
+import isEqual from "lodash.isequal";
 import { del } from "obj-case";
 import logger from "../../utils/logUtil";
```
Comment (ItsSudip) [inline integrations/Braze/browser.js:6]: Can you update the logger here with the new logger? You can refer [here](https://github.com/rudderlabs/rudder-sdk-js/pull/657). And also update the canonical name functionality similar to [this](https://github.com/rudderlabs/rudder-sdk-js/pull/685).

Comment (sauravlal15) [reply integrations/Braze/browser.js]: yeah, updated the logger.
And for the Canonical name, the PR is reverted, currently, this is not in v1-staging.

**Review by ItsSudip — APPROVED** [review]

File: integrations/Braze/browser.js [hunk 13]
```diff
@@ -124,46 +130,56 @@ class Braze {
...
-        .setDateOfBirth(
-          birthday.getUTCFullYear(),
-          birthday.getUTCMonth() + 1,
-          birthday.getUTCDate()
-        );
-    }
-
     // remove reserved keys https://www.appboy.com/documentation/Platform_Wide/#reserved-keys
     const reserved = [
-      "avatar",
```
Comment (krishna2020) [inline integrations/Braze/browser.js:166]: Are these (avatar, first_name, last_name) no more reserved keys with current braze version ?

**Review by krishna2020 — APPROVED** [review]

File: integrations/Braze/browser.js (hunk 13 above)
Comment (sauravlal15) [reply integrations/Braze/browser.js:166]: We are not using these keywords previously here and also deleting these words from traits. So removed that from the list.

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**